File: http-client/src/main/java/no/nav/familie/http/sts/StsRestClient.java
package no.nav.familie.http.sts;

import com.fasterxml.jackson.databind.ObjectMapper;
import no.nav.familie.http.client.HttpClientUtil;
import no.nav.familie.http.client.HttpRequestUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneId;
import java.util.Base64;
import java.util.concurrent.ExecutionException;

import static java.time.LocalTime.now;

public class StsRestClient {

    private static final Logger log = LoggerFactory.getLogger(StsRestClient.class);
    private final HttpClient client;
    private final URI stsUrl;
    private final String stsUsername;
    private final String stsPassword;
    private ObjectMapper mapper;
    private AccessTokenResponse cachedToken;

    public StsRestClient(ObjectMapper mapper, URI stsUrl, String stsUsername, String stsPassword) {
        this.mapper = mapper;
        this.client = HttpClientUtil.create();
        this.stsUrl = stsUrl;
        this.stsUsername = stsUsername;
        this.stsPassword = stsPassword;
    }

    private static String basicAuth(String username, String password) {
        return "Basic " + Base64.getEncoder().encodeToString((username + ":" + password).getBytes());
    }

    private boolean isTokenValid() {
        if (cachedToken == null) {
            return false;
        }

        log.debug("Tokenet løper ut: {}. Tiden nå er: {}", Instant.ofEpochMilli(cachedToken.getExpires_in()).atZone(ZoneId.systemDefault()).toLocalTime(), now(ZoneId.systemDefault()));
        return Instant.ofEpochMilli(cachedToken.getExpires_in())
            .atZone(ZoneId.systemDefault())
            .toLocalTime()
            .minusMinutes(15)
            .isAfter(now(ZoneId.systemDefault()));
    }

    public String getSystemOIDCToken() {
        if (isTokenValid()) {
            log.debug("Henter token fra cache");
            return cachedToken.getAccess_token();
        }

        log.debug("Henter token fra STS");
        HttpRequest request = HttpRequestUtil.createRequest(basicAuth(stsUsername, stsPassword))
            .uri(stsUrl)
            .header("Content-Type", "application/json")
            .timeout(Duration.ofSeconds(30))
            .build();

        AccessTokenResponse accessTokenResponse;
        try {
            accessTokenResponse = client
                .sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body)
                .thenApply(this::håndterRespons)
                .get();
        } catch (InterruptedException | ExecutionException e) {
            throw new StsAccessTokenFeilException("Feil i tilkobling", e);
        }

        if (accessTokenResponse != null) {
            this.cachedToken = accessTokenResponse;
            return accessTokenResponse.getAccess_token();
        } else {
            throw new StsAccessTokenFeilException("Manglende token");
        }
    }

    private AccessTokenResponse håndterRespons(String it) {
        try {
            return mapper.readValue(it, AccessTokenResponse.class);
        } catch (IOException e) {
            throw new StsAccessTokenFeilException("Parsing av respons feilet", e);
        }
    }
}

File: http-client/src/main/java/no/nav/familie/http/sts/StsRestClient.kt
package no.nav.familie.http.sts

import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.familie.http.client.HttpClientUtil
import no.nav.familie.http.client.HttpRequestUtil
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.*
import java.util.concurrent.ExecutionException

@Component
class StsRestClient(private val mapper: ObjectMapper,
                    @Value("\${STS_URL}") private val stsUrl: URI,
                    @Value("\${CREDENTIAL_USERNAME}") private val stsUsername: String,
                    @Value("\${CREDENTIAL_PASSWORD}") private val stsPassword: String,
                    @Value("\${STS_APIKEY:#{null}}") private val stsApiKey: String? = null) {

    private val client: HttpClient = HttpClientUtil.create()

    private var cachedToken: AccessTokenResponse? = null
    private var refreshCachedTokenTidspunkt = LocalDateTime.now()

    private val isTokenValid: Boolean
        get() {
            if (cachedToken == null) {
                return false
            }
            log.debug("Skal refreshe token: {}. Tiden nå er: {}",
                      refreshCachedTokenTidspunkt,
                      LocalTime.now())

            return refreshCachedTokenTidspunkt.isAfter(LocalDateTime.now())
        }

    val systemOIDCToken: String
        get() {
            if (isTokenValid) {
                log.debug("Henter token fra cache")
                return cachedToken!!.access_token
            }
            log.debug("Henter token fra STS")
            val request =
                    HttpRequestUtil.createRequest(basicAuth(stsUsername, stsPassword))
                            .uri(stsUrl)
                            .header("Content-Type", "application/json")
                            .POST(HttpRequest.BodyPublishers.noBody())
                            .timeout(Duration.ofSeconds(30)).apply {
                                if (!stsApiKey.isNullOrEmpty()) {
                                    header("x-nav-apiKey", stsApiKey)
                                }
                            }.build()

            val accessTokenResponse = try {
                client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                        .thenApply { obj: HttpResponse<String?> -> obj.body() }
                        .thenApply { it: String? ->
                            håndterRespons(it)
                        }
                        .get()
            } catch (e: InterruptedException) {
                throw StsAccessTokenFeilException("Feil i tilkobling", e)
            } catch (e: ExecutionException) {
                throw StsAccessTokenFeilException("Feil i tilkobling", e)
            }
            if (accessTokenResponse != null) {
                cachedToken = accessTokenResponse
                refreshCachedTokenTidspunkt = LocalDateTime.now()
                        .plusSeconds(accessTokenResponse.expires_in)
                        .minusSeconds(accessTokenResponse.expires_in / 4) // Trekker av 1/4. Refresher etter 3/4 av levetiden
                return accessTokenResponse.access_token
            }
            throw StsAccessTokenFeilException("Manglende token")
        }

    private fun håndterRespons(it: String?): AccessTokenResponse {
        return try {
            mapper.readValue(it, AccessTokenResponse::class.java)
        } catch (e: IOException) {
            throw StsAccessTokenFeilException("Parsing av respons feilet", e)
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger(StsRestClient::class.java)
        private fun basicAuth(username: String, password: String): String {
            return "Basic " + Base64.getEncoder().encodeToString("$username:$password".toByteArray())
        }
    }

}
